e2e/utils/wait-utils.ts

```diff
@@ -0,0 +1,137 @@
+ */
+export async function waitUntilMatchPageTitle(page: Page, title: string, maxTimeOut: number = 30000) {
+  return page.waitForFunction(t => {
+    const actualTitle = document.title;
+    return actualTitle.includes(t);
+  }, {timeout: maxTimeOut}, title);
+}
+
+/**
+ * Look for visible texts on the page.
+ * @param {string} texts Partial texts to look for
+ */
+export async function waitUntilPageContainsText(page: Page, texts: string, maxTimeOut: number = 30000) {
+  return page.waitForFunction(matchText => {
+    const bodyText = document.querySelector('body').innerText;
+    return bodyText.includes(matchText)
+  }, {timeout: maxTimeOut}, texts);
+}
+
+/**
+ * Summary: Helper functions when directly dealing with CSS selectors.
+ */
+
+/**
+ * Wait for the element found from the selector has a particular attribute value pair
+ * @param {string} cssSelector - The selector for the element on the page
+ * @param {string} attribute - The attribute name
+ * @param {string} value - The attribute value to match
+ */
+export async function waitUntilContainsAttributeValue(page: Page, cssSelector, attribute, value, maxTimeOut: number = 30000) {
+  return page.waitForFunction((selector, attributeName, attributeValue) => {
+    const element = document.querySelector(selector);
+    return element.attributes[attributeName] && element.attributes[attributeName].value === attributeValue;
+  }, {timeout: maxTimeOut}, cssSelector, attribute, value);
+}
+
+
+/**
+ * Wait for the elements count to be greater than a particular value
+ * @param {string} cssSelector - The css selector for the element
+ * @param {number} greaterThanCount - The expected number of elements
+ */
+export async function waitForNumberElementsIsBiggerThan(page: Page, cssSelector: string, greaterThanCount: number, maxTimeOut: number = 30000) {
+  return page.waitForFunction((selector, count) => {
+    return document.querySelectorAll(selector).length > count;
+  }, {timeout: maxTimeOut}, cssSelector, greaterThanCount);
+}
+
+
+/**
+ * Wait for the elements count to be a particular value
+ * @param {string} cssSelector - The css selector for the element
+ * @param {number} expectedCount - The expected number of elements
+ */
+export async function waitForNumberElements(page: Page, cssSelector: string, expectedCount: number, maxTimeOut: number = 30000) {
+  return page.waitForFunction((selector, count) => {
+    return document.querySelectorAll(selector).length === count;
+  }, {timeout: maxTimeOut}, cssSelector, expectedCount);
+}
+
+/**
+ * Is there a element and visible located by CSS selector?
+ *
+ * @param page Puppeteer.Page
+ * @param selector CSS selector
+ */
+export async function visible(page: Page, cssSelector: string) {
+  return page.evaluate(css => {
+    const element = document.querySelector(css);
+    return !!(element && (element.offsetWidth || element.offsetHeight || element.getClientRects().length))
+  }, cssSelector)
+}
+
+/**
+ * Wait for exact text to match.
```

Comment:
if you're using `includes()`, then it's not an exact match IIUC. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes

It looks like one branch uses `includes` and the other uses `===`. There might be a reason to do that, but it wasn't obvious.

Reply:
Looks better, though I like `waitUntil` more than `waitFor` for readability.